### DISCUSSION TITLE
请问Xray(E+F+H+A)+Caddy(N) /2_caddy中为什么有2段转发到127.0.0.1:8443

### DISCUSSION THREAD

**Discussion by CharlesWou** [lxhao61/integrated-examples#228] [Q&A] [answered]
>           {
>             "match": [{
>               "tls": {}
>             }],
>             "handle": [{
>               "handler": "proxy",
>               "upstreams": [{
>                 "dial": ["127.0.0.1:8443"] //转给 HTTP/3 server 本地监听端口
>               }],
>               "proxy_protocol": "v1" //启用 PROXY protocol 发送。v1 或 v2 表示 PROXY protocol 版本。
>             }]
>           }]
>         },
>         "udppy": {
>           "listen": ["udp/:443"],
>           "routes": [{
>             "handle": [{
>               "handler": "proxy",
>               "upstreams": [{
>                 "dial": ["udp/127.0.0.1:8443"] //转给 HTTP/3 server 本地监听端口
>               }]
>             }]
>           }]
>         } //定向 UDP 转发配置

1. 请问为什么会有这两段相同的转发？上面一段是定义TCP的，下面一段是定义UDP的？
2. 定向UDP转发是什么意思？
3. vless+xhttp+tls在走h2协议时，是走上面的转发去http中的srvh3进行处理？
4. 另外，我只用vless-vison-tls及vless-xhttp-tls这两个协议，所以我想简化下配置：layer4没做sni分流，就两个server（fwd_tcp及fwd_udp），收到tcp的直接去xray，收到udp的h3（xhttp）去127.0.0.1:8443转为h2后再去xray。那请问如果收到tcp的h2(xhttp)应该在xray做fallbacks处理还是在caddy处理比较合适？谢谢！

**Comment by lxhao61** [comment, marked as answer]
1、是的。一个 SNI 分流(基于域名的 TCP 443 转发)，一个定向 UDP 443 转发。
2、就是将流量从某 UDP 端口固定转给某 UDP 端口。
3、不是，走 h3 协议时才是。
4、不太明白你的简化办法，你可按照如下简化办法实现你的需求：
1)、Xray 中删除 Trojan+RAW+TLS 配置、VLESS+mKCP+seed 配置。
2)、Caddy 中删除分流 Trojan+RAW+TLS 配置、NaiveProxy(forwardproxy 插件) 配置。

另外若不需 SNI 分流，可选 Xray(E+H+A)+Caddy 示例来简化，简化办法类似如上。

**Reply by CharlesWou** [reply]
明白了，非常感谢！

**Reply by CharlesWou** [reply]
我想在layer4做分流，xhttp(tcp h2)及xhhtp(udp h3)走127.0.0.1:7443由caddy处理后再扔给xray，其他tcp(vless+vison+tls)默认直接扔给xray。请问如下配置是否可行？在layer4层能实现识别h2协议吗？我看layer4文档里面可以match。

```
  "apps": {
    "layer4": {
      "servers": {
        "fwd_tcp": {
          "listen": [
            "tcp/:443"
          ],
          "routes": [
            {
              "match": [
                {
                  "http": {
                    "protocol": [
                      "http/2"
                    ]
                  }
                }
              ],
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [
                    {
                      "dial": [
                        "tcp/127.0.0.1:7443"  //将XHTTP (TCP H2) 流量转给 H2/H3 server 本地监听端口
                      ]
                    }
                  ],
                  "proxy_protocol": "v1"
                }
              ]
            },
            {
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [
                    {
                      "dial": [
                        "unix/@uds5443.sock"  //转给 VLESS+Vision+TLS 监听进程
                      ]
                    }
                  ],
                  "proxy_protocol": "v1"
                }
              ]
            }
          ]
        },
        "fwd_udp": {
          "listen": [
            "udp/:443"
          ],
          "routes": [
            {
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [
                    {
                      "dial": [
                        "udp/127.0.0.1:7443"  //将XHTTP (UDP H3) 流量转给 H2/H3 server 本地监听端口
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    }
```

**Reply by lxhao61** [reply]
你的配置有问题！VLESS+Vision+TLS 默认 HTTP/2 流量，SNI 分流无法区分。

**Reply by CharlesWou** [reply]
![image](https://github.com/user-attachments/assets/efa4439f-2c9a-4fc3-94d3-5917d10329d9)

我看了下客户端填写的传输协议，VLESS+Vision+TLS 中选的是tcp，有相关文档说他也是走的h2吗？

**Reply by lxhao61** [reply]
是的。alpn 默认为 h2 与 http/1.1。

**Reply by CharlesWou** [reply]
我刚刚又看了你提到的另外一个 示例 “ Xray(E+H+A)+Caddy”：
有个疑问，为什么不能也直接在Xray中按照path分流？回落给Caddy的@uds88.sock，处理好再一次扔回Xray，这样的目的是为了？

如果我按照你这个配置改动下，利用 VLESS+Vision+TLS 回落/分流特点先分流出 VLESS+XHTTP+TLS应用，其余回落给 Caddy 处理@uds88.sock是否可行？

> Xray 前置（监听 443 端口），利用 VLESS+Vision+TLS 回落给 Caddy 处理，Caddy 为 XHTTP 提供反向代理，

**Reply by lxhao61** [reply]
你想多了！去看官方 fallbacks 资料，仅支持路径分流 WebSocket 类流量。
